bids/variables/io.py

```diff
@@ -0,0 +1,360 @@
+            duration = img.shape[3] * img.header.get_zooms()[-1]
+        except Exception as e:
+            if scan_length is not None:
+                duration = scan_length
+            else:
+                msg = ("Unable to extract scan duration from one or more "
+                       "BOLD runs, and no scan_length argument was provided "
+                       "as a fallback. Please check that the image files are "
+                       "available, or manually specify the scan duration.")
+                raise ValueError(msg)
+
+        tr = layout.get_metadata(img_f)['RepetitionTime']
+        task = entities['task']
+
+        run = dataset.get_or_create_node(entities, image_file=img_f, task=task,
+                                         duration=duration, repetition_time=tr)
+        run_info = run.get_info()
+
+        # Process event files
+        if events:
+            _data = layout.get_events(img_f, return_type='df',
+                                      derivatives=derivatives)
+            if _data is not None:
+
+                if 'amplitude' in _data.columns:
+                    if (_data['amplitude'].astype(int) == 1).all() and \
+                            'trial_type' in _data.columns:
+                        msg = ("Column 'amplitude' with constant value 1 is "
+                               "unnecessary in event files; ignoring it.")
+                        _data = _data.drop('amplitude', axis=1)
+                    else:
+                        msg = ("Column name 'amplitude' is reserved; renaming "
+                               "it to 'amplitude_'.")
+                        _data = _data.rename(
+                            columns={'amplitude': 'amplitude_'})
+                    warnings.warn(msg)
+
+                _data = _data.replace('n/a', np.nan)  # Replace BIDS' n/a
+                _data = _data.apply(pd.to_numeric, errors='ignore')
+
+                _cols = columns or list(set(_data.columns.tolist()) -
+                                        {'onset', 'duration'})
+
+                # Construct a DataFrame for each extra column
+                for col in _cols:
+                    df = _data[['onset', 'duration']].copy()
+                    df['amplitude'] = _data[col].values
+
+                    # Add in all of the run's entities as new columns for index
+                    for entity, value in entities.items():
+                        if entity in BASE_ENTITIES:
+                            df[entity] = value
+
+                    if drop_na:
+                        df = df.dropna(subset=['amplitude'])
+
+                    if df.empty:
+                        continue
+
+                    var = SparseRunVariable(col, df, run_info, 'events')
+                    run.add_variable(var)
+
+        # Process confound files
+        if confounds:
+            sub_ents = {k: v for k, v in entities.items()
+                        if k in BASE_ENTITIES}
```

Comment:
This needs to exclude `'run'` if it wasn't part of the original entities. As it is, confounds are not being pulled in.

Reply:
```Diff
diff --git a/bids/variables/io.py b/bids/variables/io.py
index 618d450..9faaafa 100644
--- a/bids/variables/io.py
+++ b/bids/variables/io.py
@@ -127,7 +127,8 @@ def _load_time_variables(layout, dataset=None, columns=None, scan_length=None,
     # Main loop over images
     for img_f in images:
 
-        entities = layout.files[img_f].entities.copy()
+        orig_entities = layout.files[img_f].entities
+        entities = orig_entities.copy()
 
         # Run is not mandatory, but we need a default for proper indexing
         if 'run' not in entities:
@@ -202,7 +203,7 @@ def _load_time_variables(layout, dataset=None, columns=None, scan_length=None,
 
         # Process confound files
         if confounds:
-            sub_ents = {k: v for k, v in entities.items()
+            sub_ents = {k: v for k, v in orig_entities.items()
                         if k in BASE_ENTITIES}
             confound_files = layout.get(type='confounds', **sub_ents)
             for cf in confound_files:
```

Reply:
Somehow this change seems also to remove "run" from the entities returned by `get_design_matrix`.